csl/core/engine.lua

```diff
@@ -0,0 +1,967 @@
+   return t
+end
+
+function CslEngine:_a_date_month (options, date)
+   local t
+   if date.month then
+      if type(date.month) == "table" then
+         local t1 = self:_a_month(options, date.month[1])
+         local t2 = self:_a_month(options, date.month[2])
+         local sep = options['range-delimiter'] or endash
+         t = t1 .. sep .. t2
+      else
+         t = self:_a_month(options, date.month)
+      end
+   elseif date.season then
+      if type(date.season) == "table" then
+         local t1 = self:_a_season(options, date.season[1])
+         local t2 = self:_a_season(options, date.season[2])
+         local sep = options['range-delimiter'] or endash
+         t = t1 .. sep .. t2
+      else
+         t = self:_a_season(options, date.season)
+      end
+   end
+   return t
+end
+
+function CslEngine:_a_date_year (options, date)
+   local t
+   if date.year then
+      if type(date.year) == "table" then
+         local t1 = self:_a_year(options, date.year[1])
+         local t2 = self:_a_year(options, date.year[2])
+         local sep = options['range-delimiter'] or endash
+         t = t1 .. sep .. t2
+      else
+         t = self:_a_year(options, date.year)
+      end
+   end
+   return t
+end
+
+function CslEngine:_date_part (options, content, date)
+   local name = SU.required(options, "name", "cs:date-part")
+   -- FIXME TODO full date range are not implemented properly
+   -- But we need to decide how to encode them in the pseudo CSL-JSON...
+   local t
+   local callback = "_a_date_" .. name
+   if self[callback] then
+      t = self[callback](self, options, date)
+   else
+      SU.warn("CSL date part " .. name .. " not implemented yet")
+   end
+   t = self:_render_textCase(t, options)
+   t = self:_render_formatting(t, options)
+   t = self:_render_affixes(t, options)
+   return t
+end
+
+function CslEngine:_date_parts (options, content, date)
+   local output = {}
+   for _, part in ipairs(content) do
+      local t = self:_date_part(part.options, part, date)
+      if t then
+         table.insert(output, t)
+      end
+   end
+   return self:_render_delimiter(output, options.delimiter)
+end
+
+function CslEngine:_date (options, content, entry)
+   local variable = SU.required(options, "variable", "CSL number")
+   local date = entry[variable]
+   self:_addGroupVariable(variable, date)
+   if date then
+      if options.form then
+         if content[1] then
+            -- NOTE: Seems the case may occcur, to select a specific date part from the date
+            -- I'm not sure I read this correctly in the CSL documentation!
+            -- That would imply usingg a copied date object, with only the relevant part?
+            SU.warn("CSL cs:date-parts in localized date not implemented")
+         end
+         -- Use locale date format
+         content = self.locale:date(options.form)
+         options.delimiter = nil -- Not supposed to exist
+      end
+      local t = self:_date_parts(options, content, date)
```

Comment:
I overlooked the `date-parts` _attribute_ on localized dates, as in `<date variable="original-date" form="text" date-parts="year" prefix="[" suffix="]"/>`
 - “year-month-day” - (default), renders the year, month and day
 - “year-month” - renders the year and month
 - “year” - renders the year

Reply:
Fixed